salt/utils/etcd_util.py

```diff
@@ -29,54 +30,69 @@
       etcd.ca: /path/to/your/ca_cert/ca.pem # Optional
       etcd.client_key: /path/to/your/client_key/client-key.pem # Optional; requires etcd.ca and etcd.client_cert to be set
       etcd.client_cert: /path/to/your/client_cert/client.pem # Optional; requires etcd.ca and etcd.client_key to be set
+      etcd.require_v2: True # Optional; defaults to True
 
 Once configured, the client() function is passed a set of opts, and optionally,
 the name of a profile to be used.
 
 .. code-block:: python
 
     import salt.utils.etcd_utils
-    client = salt.utils.etcd_utils.client(__opts__, profile='my_etcd_config')
-
-You may also use the newer syntax and bypass the generator function.
-
-.. code-block:: python
-
-    import salt.utils.etcd_utils
-    client = salt.utils.etcd_utils.EtcdClient(__opts__, profile='my_etcd_config')
-
-It should be noted that some usages of etcd require a profile to be specified,
-rather than top-level configurations. This being the case, it is better to
-always use a named configuration profile, as shown above.
+    client = salt.utils.etcd_utils.get_conn(__opts__, profile='my_etcd_config')
 """
 
 import logging
 
-from salt.exceptions import CommandExecutionError
+import salt.utils.versions
+from salt.exceptions import SaltException
 
 try:
     import etcd
     from urllib3.exceptions import ReadTimeoutError, MaxRetryError
 
-    HAS_LIBS = True
+    HAS_ETCD_V2 = True
 except ImportError:
-    HAS_LIBS = False
+    HAS_ETCD_V2 = False
+
+try:
+    import etcd3
+
+    HAS_ETCD_V3 = True
+except ImportError:
+    HAS_ETCD_V3 = False
 
 # Set up logging
 log = logging.getLogger(__name__)
 
 
+class EtcdLibraryNotInstalled(SaltException):
+    """
+    We didn't find the required etcd library
+    """
+
+
+class Etcd3DirectoryException(SaltException):
+    """
+    We didn't find the required etcd library
+    """
+
+
 class EtcdUtilWatchTimeout(Exception):
     """
     A watch timed out without returning a result
     """
 
 
 class EtcdClient:
+    """
+    Base class for the different versions of etcd clients.
+
+    This also serves as a documentation hub for all superclasses.
+    """
+
     def __init__(
         self,
         opts,
-        profile=None,
```

Comment:
We need to properly deprecate a kwarg before removing it.

Reply:
I actually meant to leave this here, will push a fix shortly.